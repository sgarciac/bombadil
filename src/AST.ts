import ct = require('chevrotain');

/**
 * The code in this file follows a general pattern.  First, I introduce a string literal
 * to use as the "tag" in the tagged unions.  I assign that literal to a variable so that
 * we don't actually use string literals in the code.  Forcing the use of the variable 
 * allows us to catch potential errors.
 * 
 * In addition, an interface for each node type is then defined.  Generally speaking, this
 * interface, while exported, isn't actually that useful outside this file.  But there are
 * some cases where they are used (hence the export).  The main reason to include the
 * interface definition is just to explicitly state the contents of the node.
 * 
 * Finally, I include a "constructor" function for each interface.  Since these are interfaces
 * and not classes, we need to define our constructor function separately.  Now, a reasonable
 * question would be "why not uses classes and a constructor defined in the class?".  The
 * reason is because once you have a class, you can use `instanceof` on the instances and
 * this is dangerous because classes are values and if classes and instances are imported
 * from different packages, things will not line up and you'll get very confusing results.
 * Using these tagged unions and interfaces avoids this but still gives a type safe way
 * to "switch" on different node types.
 */

// Table headers. We keep the Token for reporting errors during the loading phase.
export const tableHeader: 'tableHeader' = 'tableHeader';
export interface TomlTableHeader {
    type: typeof tableHeader;
    headers: string[];
    token: ct.IToken;
}
export function tomlTableHeader(headers: string[], token: ct.IToken): TomlTableHeader {
    return { type: tableHeader, headers: headers, token: token }
};

export const tableArrayEntryHeader: 'tableArrayEntryHeader' = 'tableArrayEntryHeader';
export interface TomlTableArrayEntryHeader {
    type: typeof tableArrayEntryHeader;
    headers: string[];
    token: ct.IToken;
}
export function tomlTableArrayEntryHeader(headers: string[], token: ct.IToken): TomlTableArrayEntryHeader {
    return { type: tableArrayEntryHeader, headers: headers, token: token }
}

// Bindings
export const keyValue: 'keyValue' = 'keyValue';
export interface TomlKeyValue {
    type: typeof keyValue;
    key: string;
    value: any;
    token: ct.IToken;
}
export function tomlKeyValue(key: string, value: any, token: ct.IToken): TomlKeyValue {
    return { type: keyValue, key: key, value: value, token: token }
};

// Structures
export const inlineTable: 'inlineTable' = 'inlineTable';
export interface TomlInlineTable {
    type: typeof inlineTable;
    bindings: TomlKeyValue[];
}
export function tomlInlineTable(bindings: TomlKeyValue[]): TomlInlineTable {
    return { type: inlineTable, bindings: bindings };
}

export const arrayType: 'tomlArray' = 'tomlArray';
export interface TomlArray {
    type: typeof arrayType;
    contents: TomlValue[];
    token: ct.IToken;
}
export function tomlArray(contents: TomlValue[], token: ct.IToken) {
    return { type: arrayType, contents: contents, token: token };
}

// Atomic Values
export enum TomlAtomicValueType {
    OffsetDateTime, LocalDateTime, LocalDate, LocalTime, String, Integer, Float, Boolean
}

export interface TomlAtomicGeneric<T> {
    image: string;
    value: T;
}

// For atomic data, first we list all the node types
export const offsetDateTime: 'offsetDateTime' = 'offsetDateTime';
export const localDateTime: 'localDateTime' = 'localDateTime';
export const localDate: 'localDate' = 'localDate';
export const localTime: 'localTime' = 'localTime';

// then we list a single parameterized interface for all nodes (since they all have
// the same structure and only the type of the data changes).
export interface TomlAtomicDateTime extends TomlAtomicGeneric<Date> {
    type: typeof offsetDateTime | typeof localDateTime | typeof localDate | typeof localTime;
}

// Now, the constructor functions.
export function tomlAtomicOffsetDateTime(image: string, value: Date): TomlAtomicDateTime {
    return { type: offsetDateTime, image: image, value: value };
}
export function tomlAtomicLocalDateTime(image: string, value: Date): TomlAtomicDateTime {
    return { type: localDateTime, image: image, value: value };
}
export function tomlAtomicLocalDate(image: string, value: Date): TomlAtomicDateTime {
    return { type: localDate, image: image, value: value };
}
export function tomlAtomicLocalTime(image: string, value: Date): TomlAtomicDateTime {
    return { type: localTime, image: image, value: value };
}

export const atomicString: 'atomicString' = 'atomicString';
export interface TomlAtomicString extends TomlAtomicGeneric<string> {
    type: typeof atomicString;
}
export function tomlAtomicString(image: string, value: string): TomlAtomicString {
    return { type: atomicString, image: image, value: value };
}

export const atomicInteger: 'atomicInteger' = 'atomicInteger';
export interface TomlAtomicInteger extends TomlAtomicGeneric<number> {
    type: typeof atomicInteger;
}
export function tomlAtomicInteger(image: string, value: number): TomlAtomicInteger {
    return { type: atomicInteger, image: image, value: value };
}

export const atomicFloat: 'atomicFloat' = 'atomicFloat';
export interface TomlAtomicFloat extends TomlAtomicGeneric<number> {
    type: typeof atomicFloat;
}
export function tomlAtomicFloat(image: string, value: number): TomlAtomicFloat {
    return { type: atomicFloat, image: image, value: value };
}

export const atomicBoolean: 'atomicBoolean' = 'atomicBoolean';
export interface TomlAtomicBoolean extends TomlAtomicGeneric<boolean> {
    type: typeof atomicBoolean;
}
export function tomlAtomicBoolean(image: string, value: boolean): TomlAtomicBoolean {
    return { type: atomicBoolean, image: image, value: value };
}

// Now we define a few convenient union types to represent essentially different
// grammatic productions
export type TomlAtomicValue =
    TomlAtomicDateTime | TomlAtomicString | TomlAtomicInteger | TomlAtomicFloat | TomlAtomicBoolean;

export type TopLevelTomlDocumentEntry = (TomlKeyValue | TomlTableHeader | TomlTableArrayEntryHeader)

export type TomlValue = (
    TomlAtomicValue |
    TomlInlineTable |
    TomlArray);
