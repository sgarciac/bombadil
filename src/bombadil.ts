export * from './tables';
export * from './lexer';
export * from './parser';
export * from './AST';
