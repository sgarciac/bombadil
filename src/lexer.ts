import ct = require('chevrotain')


export class OneLineComment extends ct.Token {
    static PATTERN = /#.*/;
    static GROUP = ct.Lexer.SKIPPED;
}

// Basic identifiers
export class Identifier extends ct.Token {
    static PATTERN = /[A-Za-z0-9_-]+/;
}

// Atomic values
export class Integer extends ct.Token {
    static PATTERN = /[+-]?(\d_|_\d|\d)+/;
}

export class Float extends ct.Token {
    static PATTERN = /([+-]?(\d_|_\d|\d)+)(((\.(\d_|_\d|\d)+)([Ee]([+-])?(\d_|_\d|\d)+))|((\.(\d_|_\d|\d)+)|([Ee]([+-])?(\d_|_\d|\d)+)))/;
}

export class Booolean extends ct.Token {
    static PATTERN = /true|false/;
}

export class OffsetDateTime extends ct.Token {
    static PATTERN = /-?\d{4}-\d{2}-\d{2}(t|\s)\d{2}:\d{2}:\d{2}(\.\d+)?(z|([-+]\d{2}:\d{2}))/i
}

export class LocalDateTime extends ct.Token {
    static PATTERN = /-?\d{4}-\d{2}-\d{2}(t|\s)\d{2}:\d{2}:\d{2}(\.\d+)?/i
}

export class LocalDate extends ct.Token {
    static PATTERN = /-?\d{4}-\d{2}-\d{2}/i
}

export class LocalTime extends ct.Token {
    static PATTERN = /\d{2}:\d{2}:\d{2}(\.\d+)?/i
}

export class EndOfLine extends ct.Token {
    static PATTERN = /(\r\n|\n)+/;
    static GROUP = ct.Lexer.SKIPPED;
}

export class WhiteSpace extends ct.Token {
    static PATTERN = /[^\S\n\r]+/;
    static GROUP = ct.Lexer.SKIPPED;
}

export class OpenValue extends ct.Token {
    static PATTERN = /=/;
    static PUSH_MODE = 'value'
}

export class CloseValue extends ct.Token {
    static PATTERN = /(\r\n|\n)+/;
    static POP_MODE = true;
}

export class OpenInlineTable extends ct.Token {
    static PATTERN = /\{/;
    static PUSH_MODE = 'inline_table'
}

export class CloseInlineTable extends ct.Token {
    static PATTERN = /\}/;
    static POP_MODE = true;
}

export class OpenInlineValue extends ct.Token {
    static PATTERN = /=/;
    static PUSH_MODE = 'inline_value'
}

export class CloseInlineValue extends ct.Token {
    // hackish way to use } as end of value without consuming it
    static PATTERN = /,|(.{0}(?=}))/;
    static POP_MODE = true;
}

export class Comma extends ct.Token {
    static PATTERN = /,/;
}

export class Dot extends ct.Token {
    static PATTERN = /\./;
}

export class OpenMultiLineBasicString extends ct.Token {
    static PATTERN = /"""/;
    static PUSH_MODE = 'multi_line_basic_string';
}

export class CloseMultiLineBasicString extends ct.Token {
    static PATTERN = /"""/;
    static POP_MODE = true;
}

export class OpenBasicString extends ct.Token {
    static PATTERN = /"/;
    static PUSH_MODE = 'basic_string';
}

export class CloseBasicString extends ct.Token {
    static PATTERN = /"/;
    static POP_MODE = true;
}

export class OpenLiteralString extends ct.Token {
    static PATTERN = /'/;
    static PUSH_MODE = 'literal_string';
}

export class CloseLiteralString extends ct.Token {
    static PATTERN = /'/;
    static POP_MODE = true;
}

export class OpenMultiLineLiteralString extends ct.Token {
    static PATTERN = /'''/;
    static PUSH_MODE = 'multi_line_literal_string';
}

export class CloseMultiLineLiteralString extends ct.Token {
    static PATTERN = /'''/;
    static POP_MODE = true;
}

export class EscapedChar extends ct.Token {
    static PATTERN = /(\\b)|(\\t)|(\\n)|(\\f)|(\\")|(\\r)|(\\\\)/
}

export class EscapedUnicode extends ct.Token {
    static PATTERN = /(\\u([0-9A-Fa-f]{4}))|(\\U([0-9A-Fa-f]{8}))/
}

export class SubBasicString extends ct.Token {
    static PATTERN = /[^\\"\r\n]+/
}

export class SubMultiLineBasicString extends ct.Token {
    static PATTERN = /([^\\"]|"(?!""))+/
}

export class MultiLineIgnorableSubstring extends ct.Token {
    static PATTERN = /\\\s*(\r\n|\n)\s*/;
    static GROUP = ct.Lexer.SKIPPED;
}

export class LiteralString extends ct.Token {
    static PATTERN = /[^'\n\r]+/;
}

export class MultiLineLiteralString extends ct.Token {
    static PATTERN = /([^']|'(?!''))+/;
}

export class OpenArray extends ct.Token {
    static PATTERN = /\[/;
    static PUSH_MODE = 'array';
}

export class CloseArray extends ct.Token {
    static PATTERN = /\]/;
    static POP_MODE = true;
}

export class OpenTable extends ct.Token {
    static PATTERN = /\[/;
    static PUSH_MODE = 'table';
}

export class CloseTable extends ct.Token {
    static PATTERN = /\]((([^\S\n\r]*#.*)|[^\S\n\r]*)(\r\n|\n))/;
    static POP_MODE = true;
}

export class OpenTableArrayItem extends ct.Token {
    static PATTERN = /\[\[/;
    static PUSH_MODE = 'table_array_item';
}

export class CloseTableArrayItem extends ct.Token {
    static PATTERN = /\]\][^\S\n\r]*/;
    static POP_MODE = true;
}


var open_all_strings: ct.TokenConstructor[] = [
    OpenMultiLineBasicString,
    OpenMultiLineLiteralString,
    OpenBasicString,
    OpenLiteralString];

var atomic_literals: ct.TokenConstructor[] = [
    OffsetDateTime,
    LocalDateTime,
    LocalDate,
    LocalTime,
    Float,
    Integer,
    Booolean,
]

var open_identifier_strings: ct.TokenConstructor[] = [
    OpenBasicString,
    OpenLiteralString
]

var single_line_skipped: ct.TokenConstructor[] = [
    WhiteSpace,
    OneLineComment,
]

var all_skipped: ct.TokenConstructor[] = [
    WhiteSpace,
    EndOfLine,
    OneLineComment,
]

var modes: ct.IMultiModeLexerDefinition = {
    modes: {
        top: [
            OpenTableArrayItem,
            OpenTable,
            Identifier,
            ...open_identifier_strings,
            ...all_skipped,
            OpenValue,
        ],
        value: [
            ...open_all_strings,
            ...atomic_literals,
            ...single_line_skipped,
            OpenArray,
            OpenInlineTable,
            CloseValue
        ],
        table: [
            Identifier,
            ...open_identifier_strings,
            Dot,
            WhiteSpace,
            CloseTable
        ]
        ,
        table_array_item: [
            Identifier,
            ...open_identifier_strings,
            Dot,
            WhiteSpace,
            CloseTableArrayItem
        ]
        ,
        array: [
            ...atomic_literals,
            ...all_skipped,
            ...open_all_strings,
            Comma,
            OpenArray,
            OpenInlineTable,
            CloseArray
        ],
        inline_table: [
            Identifier,
            ...open_identifier_strings,
            ...single_line_skipped,
            OpenInlineValue,
            CloseInlineTable
        ],
        inline_value: [
            ...open_all_strings,
            ...atomic_literals,
            ...single_line_skipped,
            OpenArray,
            OpenInlineTable,
            CloseInlineValue
        ],
        basic_string: [
            CloseBasicString,
            EscapedChar,
            EscapedUnicode,
            SubBasicString],
        multi_line_basic_string: [
            CloseMultiLineBasicString,
            EscapedChar,
            EscapedUnicode,
            MultiLineIgnorableSubstring,
            SubMultiLineBasicString],
        literal_string: [LiteralString,
            CloseLiteralString],
        multi_line_literal_string: [
            MultiLineLiteralString,
            CloseMultiLineLiteralString]
    },
    defaultMode: 'top'
}

export var tomlLexer = new ct.Lexer(modes);

export var allTokens = [];
for (let symbol in this) {
    if (this[symbol].tokenType) {
        allTokens.push(this[symbol]);
    }
}



