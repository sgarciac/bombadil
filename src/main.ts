import toml = require('./bombadil')

import fs = require('fs')

var input = fs.readFileSync('example.toml','utf8');

var reader = new toml.TomlReader();
reader.readToml(input, true);

function bombadilToTomlTest(input){
    if (input instanceof toml.TomlAtomicValue){
        return input.value;
    } else if (input instanceof Array){
        return input.map(bombadilToTomlTest);
    } else {
        let newObj = {}
        for(let property in input){
            newObj[property] = bombadilToTomlTest(input[property]);
        }
        return newObj;
    }
}

console.log(JSON.stringify(bombadilToTomlTest(reader.result), null, 2));
//console.log(JSON.stringify(reader.result, null, 2))