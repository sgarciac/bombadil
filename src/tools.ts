function fromCodePoint(codePoint) {
    let codeUnits = [];
    let highSurrogate;
    let lowSurrogate;
    if (codePoint > 0x10FFFF) {
        throw RangeError('Invalid code point: ' + codePoint);
    }
    if (codePoint <= 0xFFFF) {
        codeUnits.push(codePoint);
    } else {
        codePoint -= 0x10000;
        highSurrogate = (codePoint >> 10) + 0xD800;
        lowSurrogate = (codePoint % 0x400) + 0xDC00;
        codeUnits.push(highSurrogate, lowSurrogate);
    }
    return String.fromCharCode.apply(null, codeUnits);
}

export function unicodeToString(unicode): string {
    let size = (unicode[1] == 'u') ? 4 : 6;
    let codeString = unicode.substr(2, 1 + size);
    return fromCodePoint([parseInt(codeString, 16)]);
}

export function cleanNumberImage(image: string): string {
    return image.replace('_', '');
}

export function startsWithEOL(string): boolean {
    return string.match(/^(\n|\r\n)/);
}

export function trimWhiteSpacePrefix(string) {
    if (startsWithEOL(string)) {
        return string.replace(/^[\s\uFEFF\xA0]+/, '');
    } else {
        return string;
    }
}

export function escapedToString(escaped: string): string {
    switch (escaped) {
        case '\\n':
            return '\n';
        case '\\r':
            return '\r';
        case '\\\\':
            return '\\'
        case '\\\"':
            return '\"';
        case '\\b':
            return '\b';
        case '\\t':
            return '\t';
        case '\\f':
            return '\f';
        default:
            throw 'unrecognised escaped char';
    }
}

