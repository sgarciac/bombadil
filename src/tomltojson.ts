import { TomlReader } from "./bombadil";
import * as util from "util";
import * as fs from "fs";
const input = fs.readFileSync("/dev/stdin", "utf8");

const reader = new TomlReader();
reader.readToml(input, true);
if (reader.result) {
  console.log(util.inspect(reader.result, { showHidden: false, depth: null }));
} else {
  console.log(reader.errors);
}
