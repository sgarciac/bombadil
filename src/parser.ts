import ct = require('chevrotain');
import * as l from './lexer';
import * as tools from './tools';
import * as ast from './AST';

export class TomlParser extends ct.Parser {

    public documentRule = this.RULE<ast.TopLevelTomlDocumentEntry[]>('documentRule', () => {
        let documentEntries: ast.TopLevelTomlDocumentEntry[] = [];
        this.MANY(() => {
            this.OR(
                [
                    { ALT: () => { documentEntries.push(this.SUBRULE(this.tableHeaderRule)) } },
                    { ALT: () => { documentEntries.push(this.SUBRULE(this.tableArrayEntryHeaderRule)) } },
                    { ALT: () => { documentEntries.push(this.SUBRULE(this.keysValueRule)) } },
                    { ALT: () => { this.CONSUME(l.EndOfLine) } }
                ]
            );
        });
        return documentEntries;
    });

    keysValueRule = this.RULE('keysValueRule', () => {
        let keys: string[] = [];
        this.AT_LEAST_ONE_SEP({
            SEP: l.Dot, DEF: () => {
                keys.push(this.SUBRULE(this.identifierRule));
            }
        });

        let equals = this.CONSUME(l.OpenValue);
        let value = this.SUBRULE(this.valueRule);
        this.OR([
            { ALT: () => this.CONSUME(l.CloseValue) },
            { ALT: () => this.CONSUME(ct.EOF) }
        ]);
        //this.CONSUME(l.CloseValue);
        return ast.tomlKeysValue(keys, value, equals);
    })

    valueRule = this.RULE('valueRule', () => {
        let value: any;
        this.OR([
            // Atomic values
            { ALT: () => { let image = this.CONSUME(l.Float).image; value = ast.tomlAtomicFloat(image, tools.parseFloat(image)) } },
            { ALT: () => { let image = this.CONSUME(l.TomlInfinity).image; value = ast.tomlAtomicInfinity(image, tools.parseInfinity(image)) } },
            { ALT: () => { let image = this.CONSUME(l.TomlNotANumber).image; value = ast.tomlAtomicNotANumber(image, tools.parseNotANumber(image)) } },
            { ALT: () => { let image = this.CONSUME(l.Integer).image; value = ast.tomlAtomicInteger(image, tools.parseDecInteger(image)) } },
            { ALT: () => { let image = this.CONSUME(l.BinaryInteger).image; value = ast.tomlAtomicInteger(image, tools.parseBinaryInteger(image)) } },
            { ALT: () => { let image = this.CONSUME(l.OctalInteger).image; value = ast.tomlAtomicInteger(image, tools.parseOctalInteger(image)) } },
            { ALT: () => { let image = this.CONSUME(l.HexInteger).image; value = ast.tomlAtomicInteger(image, tools.parseHexInteger(image)) } },
            { ALT: () => { let image = this.CONSUME(l.Booolean).image; value = ast.tomlAtomicBoolean(image, tools.parseBoolean(image)) } },
            { ALT: () => { let image = this.CONSUME(l.OffsetDateTime).image; value = ast.tomlAtomicOffsetDateTime(image, tools.parseOffetDateTime(image)) } },
            { ALT: () => { let image = this.CONSUME(l.LocalDateTime).image; value = ast.tomlAtomicLocalDateTime(image, tools.parseLocalDateTime(image)) } },
            { ALT: () => { let image = this.CONSUME(l.LocalDate).image; value = ast.tomlAtomicLocalDate(image, tools.parseLocalDate(image)) } },
            { ALT: () => { let image = this.CONSUME(l.LocalTime).image; value = ast.tomlAtomicLocalTime(image, tools.parseLocalTime(image)) } },

            // structures
            { ALT: () => { value = this.SUBRULE(this.arrayRule) } },
            { ALT: () => { value = this.SUBRULE(this.inlineTableRule) } },
            //Strings
            { ALT: () => { value = this.SUBRULE(this.literalStringRule) } },
            { ALT: () => { value = this.SUBRULE(this.multiLineBasicStringRule) } },
            { ALT: () => { value = this.SUBRULE(this.multiLineLiteralStringRule) } },
            { ALT: () => { value = this.SUBRULE(this.basicStringRule) } }
        ]);
        return value;
    })

    // not using MANY_SEP because of trailing commas.
    arrayRule = this.RULE('arrayRule', () => {
        let values: any[] = [];
        let token = this.CONSUME(l.OpenArray);
        this.OPTION(() => {
            values.push(this.SUBRULE(this.valueRule));
            this.MANY(() => {
                this.CONSUME(l.Comma);
                values.push(this.SUBRULE2(this.valueRule));
            })
            this.MANY2(() => { this.CONSUME2(l.Comma) });
        })
        this.CONSUME(l.CloseArray);
        return ast.tomlArray(values, token);
    })

    inlineTableRule = this.RULE('inlineTableRule', () => {
        let bindings: ast.TomlKeysValue[] = [];
        this.CONSUME(l.OpenInlineTable);
        this.MANY(() => {
            let keys: string[] = [];
            this.AT_LEAST_ONE_SEP({
                SEP: l.Dot, DEF: () => {
                    keys.push(this.SUBRULE(this.identifierRule));
                }
            });
            let equals = this.CONSUME(l.OpenInlineValue);
            let value = this.SUBRULE(this.valueRule);
            this.CONSUME(l.CloseInlineValue);
            bindings.push(ast.tomlKeysValue(keys, value, equals));
        });
        this.CONSUME(l.CloseInlineTable);
        return ast.tomlInlineTable(bindings);
    })


    tableHeaderRule = this.RULE('tableHeaderRule', (): ast.TomlTableHeader => {
        let headers: string[] = [];
        let open_table = this.CONSUME(l.OpenTable);
        this.AT_LEAST_ONE_SEP({
            SEP: l.Dot, DEF: () => {
                headers.push(this.SUBRULE(this.identifierRule));
            }
        });
        this.CONSUME(l.CloseTable);
        this.OR([
            { ALT: () => this.CONSUME(l.EndOfLine) },
            { ALT: () => this.CONSUME(ct.EOF) }
        ]);
        return ast.tomlTableHeader(headers, open_table);
    });

    tableArrayEntryHeaderRule = this.RULE('tableArrayEntryHeaderRule', (): ast.TomlTableArrayEntryHeader => {
        let headers: string[] = [];
        let open_table = this.CONSUME(l.OpenTableArrayItem);
        this.AT_LEAST_ONE_SEP({
            SEP: l.Dot, DEF: () => {
                headers.push(this.SUBRULE(this.identifierRule));
            }
        });
        this.CONSUME(l.CloseTableArrayItem)
        this.OR([
            { ALT: () => this.CONSUME(l.EndOfLine) },
            { ALT: () => this.CONSUME(ct.EOF) }
        ]);
        return ast.tomlTableArrayEntryHeader(headers, open_table);
    });

    basicStringRule = this.RULE('basicStringRule', () => {
        let basicString: string = '';
        let fullImage: string = '';
        this.CONSUME(l.OpenBasicString);
        this.MANY(() => {
            this.OR([
                { ALT: () => { let image = this.CONSUME(l.EscapedChar).image; basicString += tools.parseEscapedCharacter(image); fullImage += image } },
                {
                    ALT: () => {
                        let token = this.CONSUME(l.EscapedUnicode);
                        try {
                            basicString += tools.parseEscapedUnicode(token.image);
                        } catch {
                            throw { message: 'Bad codepoint', token: token };
                        }
                        fullImage += token.image
                    }
                },
                { ALT: () => { let image = this.CONSUME(l.SubBasicString).image; basicString += image; fullImage += image } }
            ])
        });
        this.CONSUME(l.CloseBasicString);
        return ast.tomlAtomicString(fullImage, basicString);

    });

    literalStringRule = this.RULE('literalStringRule', (): ast.TomlAtomicString => {
        let literalString: string = '';
        this.CONSUME(l.OpenLiteralString);
        this.OPTION(() => { literalString += this.CONSUME(l.LiteralString).image });
        this.CONSUME(l.CloseLiteralString);
        return ast.tomlAtomicString(literalString, literalString);
    });

    multiLineBasicStringRule = this.RULE('multiLineBasicStringRule', () => {
        let multiLineString: string = '';
        let fullImage: string = '';
        this.CONSUME(l.OpenMultiLineBasicString);
        this.MANY(() => {
            this.OR([
                { ALT: () => { let image = this.CONSUME(l.EscapedChar).image; multiLineString += tools.parseEscapedCharacter(image); fullImage += image } },
                {
                    ALT: () => {
                        let token = this.CONSUME(l.EscapedUnicode);
                        try {
                            multiLineString += tools.parseEscapedUnicode(token.image);
                        } catch {
                            throw { message: 'Bad codepoint', token: token };
                        }
                        fullImage += token.image
                    }
                },
                { ALT: () => { let image = this.CONSUME(l.SubMultiLineBasicString).image; multiLineString += image; fullImage += image } }
            ])
        });
        this.CONSUME(l.CloseMultiLineBasicString);
        return ast.tomlAtomicString(fullImage, tools.trimWhiteSpacePrefix(multiLineString));
    });

    multiLineLiteralStringRule = this.RULE('multiLineLiteralStringRule', () => {
        let value: string;
        this.CONSUME(l.OpenMultiLineLiteralString);
        this.OPTION(() => {
            value = this.CONSUME(l.MultiLineLiteralString).image;
        });
        this.CONSUME(l.CloseMultiLineLiteralString);
        return ast.tomlAtomicString(value, tools.trimWhiteSpacePrefix(value));
    });

    //bareKeyworkRule = this.RULE('bareKeyword', () => {this.CONSUME(l.Identifier)});
    identifierRule = this.RULE('identifierRule', (): string => {
        let id: string;
        this.OR([
            { ALT: () => { id = this.CONSUME(l.Identifier).image } },
            { ALT: () => { id = this.SUBRULE(this.literalStringRule).value } },
            { ALT: () => { id = this.SUBRULE(this.basicStringRule).value } }
        ]);
        return id;
    });

    constructor(tokenTypes: ct.TokenVocabulary) {
        super(tokenTypes, { outputCst: false });
        this.performSelfAnalysis();
    }
}
