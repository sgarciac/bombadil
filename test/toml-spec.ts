import { readFileSync } from 'fs';
import * as path from 'path';
import { TomlReader } from '../src/bombadil';
import { safeLoad } from 'js-yaml';

function readSample(name: string): string {
    return readFileSync(path.join(__dirname, 'toml-spec-samples', name)).toString();
}

function compare(toml: string) {
    let input = readSample(toml + '.toml');
    let baseline = safeLoad(readSample(toml + '.yaml'));
    let reader = new TomlReader();
    reader.readToml(input);
    expect(reader.result).toEqual(baseline);
}

test('Example TOML', () => {
    compare('example');
});

test('Fruit TOML', () => {
    compare('fruit');
});

test('Hard Example TOML', () => {
    compare('hard_example');
});

test('Hard Example TOML', () => {
    let input = readSample('hard_example_unicode.toml');
    let reader = new TomlReader();
    reader.readToml(input);
    expect(reader.result).toMatchSnapshot();
});