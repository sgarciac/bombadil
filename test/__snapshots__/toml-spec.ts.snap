// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hard Example TOML 1`] = `
Object {
  "the": Object {
    "hard": Object {
      "another_test_string": "§á₥è ƭλïñϱ, βúƭ ωïƭλ á ƨƭřïñϱ #",
      "harder_test_string": " Âñδ ωλèñ \\"'ƨ ářè ïñ ƭλè ƨƭřïñϱ, áℓôñϱ ωïƭλ # \\"",
      "test_array": Array [
        "] ",
        " # ",
      ],
      "test_array2": Array [
        "Tèƨƭ #11 ]ƥřôƲèδ ƭλáƭ",
        "Éжƥèřï₥èñƭ #9 ωáƨ á ƨúççèƨƨ",
      ],
      "βïƭ#": Object {
        "multi_line_array": Array [
          "]",
        ],
        "ωλáƭ?": "Ýôú δôñ'ƭ ƭλïñƙ ƨô₥è úƨèř ωôñ'ƭ δô ƭλáƭ?",
      },
    },
    "test_string": "Ýôú'ℓℓ λáƭè ₥è áƒƭèř ƭλïƨ - #",
  },
}
`;
